src/D2L.CodeStyle.Analyzers/Helpers/ConstantAnalyzer.cs

```diff
@@ -0,0 +1,151 @@
+﻿using System.Collections.Immutable;
+using D2L.CodeStyle.Analyzers.Extensions;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.CodeAnalysis.CSharp.Syntax;
+using Microsoft.CodeAnalysis.Diagnostics;
+
+namespace D2L.CodeStyle.Analyzers.Helpers {
+	[DiagnosticAnalyzer( LanguageNames.CSharp )]
+	public sealed class ConstantAnalyzer : DiagnosticAnalyzer {
+		private const string AttributeName = "D2L.CodeStyle.Annotations.Contract.ConstantAttribute";
+
+		public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics =>
+			ImmutableArray.Create(
+				Diagnostics.NonConstantPassedToConstantParameter,
+				Diagnostics.InvalidConstantType
+			);
+
+		public override void Initialize( AnalysisContext context ) {
+			context.EnableConcurrentExecution();
+			context.ConfigureGeneratedCodeAnalysis( GeneratedCodeAnalysisFlags.Analyze | GeneratedCodeAnalysisFlags.ReportDiagnostics );
+			context.RegisterCompilationStartAction( CompilationStart );
+		}
+
+		public static void CompilationStart(
+			CompilationStartAnalysisContext context
+		) {
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeParameters(
+					ctx,
+					(ParameterListSyntax)ctx.Node
+				),
+				SyntaxKind.ParameterList
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					( (InvocationExpressionSyntax)ctx.Node ).ArgumentList
+				),
+				SyntaxKind.InvocationExpression
+			);
+
+			context.RegisterSyntaxNodeAction(
+				ctx => AnalyzeArguments(
+					ctx,
+					( (ObjectCreationExpressionSyntax)ctx.Node ).ArgumentList
+				),
+				SyntaxKind.ObjectCreationExpression
+			);
+		}
+
+		private static void AnalyzeParameters(
+			SyntaxNodeAnalysisContext context,
+			ParameterListSyntax parameterList
+		) {
+			SeparatedSyntaxList<ParameterSyntax> parameters = parameterList.Parameters;
+			foreach( ParameterSyntax parameterSyntax in parameters ) {
+				// We need the symbol
+				IParameterSymbol parameter = context.SemanticModel.GetDeclaredSymbol( parameterSyntax );
+
+				// Parameter is somehow null, so do nothing
+				if( parameter == null ) {
+					continue;
+				}
+
+				// Parameter is not [Constant], so do nothing
+				if( !parameter.HasAttribute( AttributeName ) ) {

```

Comment:
Check `StatelessFuncAnalyzer` for an easy reference of doing this with symbols.

https://github.com/Brightspace/D2L.CodeStyle/blob/c9bacec35a74b8568721dd25a9dad187a5d1c826/src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs#L28

https://github.com/Brightspace/D2L.CodeStyle/blob/c9bacec35a74b8568721dd25a9dad187a5d1c826/src/D2L.CodeStyle.Analyzers/Immutability/StatelessFuncAnalyzer.cs#L64